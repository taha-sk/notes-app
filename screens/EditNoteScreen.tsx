import { StackScreenProps } from '@react-navigation/stack';
import React from 'react';
import { ScrollView, StyleSheet, TextInput, View, Text, Pressable } from 'react-native';
import { NoteItem, RootStackParamList } from '../types';
import AsyncStorage from '@react-native-async-storage/async-storage';
import { AntDesign } from '@expo/vector-icons';
import { v4 as uuidv4 } from 'uuid'; 

export default function EditNoteScreen({navigation, route} : StackScreenProps<RootStackParamList, 'EditNote'>) {
  
  const { noteId } = route.params;

  const init: NoteItem[] =[];
  const [notes, setNotes] = React.useState(init);

  const [note, setNote] = React.useState('');
  const [err, setErr] = React.useState('');

  const initializeData = async () => {
    try {
      const jsonValue = await AsyncStorage.getItem('@notes');
      if(jsonValue != null){
        const storedNotes: NoteItem[] = JSON.parse(jsonValue);
        setNotes(storedNotes);
        if(noteId.length > 0){
          const storedItem = storedNotes.find(notes => notes.id === noteId);
          if(storedItem){
            setNote(storedItem.note);
          }
        }
      }
    } catch(e) {
      const errorMessage = (e as Error).message
      setErr('Error occured during data initialization. Please restart the application and try again. Error detail: ' + errorMessage);
    }
  };

  const storeData = async (value: NoteItem[]) => {
    try {
      const jsonValue = JSON.stringify(value)
      await AsyncStorage.setItem('@notes', jsonValue)
    } catch (e) {
      const errorMessage = (e as Error).message
      setErr('Error occured during data save. Please restart the application and try again. Error detail: ' + errorMessage);
    }
  };

  const saveNoteAndExit = () => {
    if(note.length > 0){
      if(noteId.length > 0){
        const updatedNotes = notes.map((val) => val.id === noteId ? {id: val.id, note: note} : val);
        storeData(updatedNotes);
      }else{
        const noteItem: NoteItem = { id: uuidv4(), note: note};
        storeData([noteItem, ...notes]);
      }
    }
    navigation.goBack();
  };

  const deleteNote = () => {
    if(noteId.length > 0){
      const updatedNotes = notes.filter((val) => val.id !== noteId);
      storeData(updatedNotes);
    }
    navigation.goBack();
  };

  React.useEffect(() => {initializeData()}, []);

  return (
    <View style={styles.container}>
      <ScrollView>
        <TextInput
          style={{height: 275, fontSize: 20}}
          multiline
          placeholder="Type here to add your note!"
          onChangeText={text => setNote(text)}
          defaultValue={note}
        />
      </ScrollView>
      {err.length > 0 && <Text>{err}</Text>}
      <View style={styles.footer}>
        <Pressable onPress={() => deleteNote()}>
          <AntDesign name="delete" size={30} color="#dec4d6" />
        </Pressable>        
        <Pressable onPress={() => saveNoteAndExit()}>
          <AntDesign name="save" size={30} color="#37667e" />
        </Pressable>
      </View>
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#f0f1f6',
    padding: 20
  },
  footer: {
    flexDirection: 'row',
    justifyContent: 'space-evenly'
  }
});
