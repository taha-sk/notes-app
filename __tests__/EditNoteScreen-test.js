import * as React from 'react';
import renderer from 'react-test-renderer';
import { render, fireEvent } from '@testing-library/react-native';

import EditNoteScreen from '../screens/EditNoteScreen';

describe('<EditNoteScreen />', () => {

  it('renders correctly', () => {
    const EditNoteScreenProps = {
      navigation: {goBack: jest.fn()},
      route: {params: { noteId: "1"}}
    };
    const tree = renderer.create(<EditNoteScreen {...EditNoteScreenProps}/>).toJSON();
    expect(tree).toMatchSnapshot();
  });  

  it('has 2 child', () => {
    const EditNoteScreenProps = {
      navigation: {goBack: jest.fn()},
      route: {params: { noteId: "1"}}
    };
    const tree = renderer.create(<EditNoteScreen {...EditNoteScreenProps}/>).toJSON();
    expect(tree.children.length).toBe(2);
  });

  it('saves correctly', () => {
    const EditNoteScreenProps = {
      navigation: {goBack: jest.fn()},
      route: {params: { noteId: "1"}}
    };
    const { getByPlaceholderText, getByTestId } = render(<EditNoteScreen {...EditNoteScreenProps} />);
    fireEvent.changeText(
      getByPlaceholderText('Type here to add your note!'),
      'test note'
    );
     const saveBtn = getByTestId('save-btn');
     fireEvent.press(saveBtn);
    expect(EditNoteScreenProps.navigation.goBack).toHaveBeenCalledTimes(1);
  });

  it('deletes correctly', () => {
    const EditNoteScreenProps = {
      navigation: {goBack: jest.fn()},
      route: {params: { noteId: "1"}}
    };
    const { getByTestId } = render(<EditNoteScreen {...EditNoteScreenProps} />);
     const deleteBtn = getByTestId('delete-btn');
     fireEvent.press(deleteBtn);
    expect(EditNoteScreenProps.navigation.goBack).toHaveBeenCalledTimes(1);
  });  

});
