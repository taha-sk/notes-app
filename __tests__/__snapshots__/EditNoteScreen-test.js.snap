// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EditNoteScreen /> renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#f0f1f6",
      "flex": 1,
      "padding": 20,
    }
  }
>
  <RCTScrollView>
    <View>
      <TextInput
        allowFontScaling={true}
        defaultValue=""
        multiline={true}
        onChangeText={[Function]}
        placeholder="Type here to add your note!"
        rejectResponderTermination={true}
        style={
          Object {
            "fontSize": 20,
            "height": 275,
          }
        }
        underlineColorAndroid="transparent"
      />
    </View>
  </RCTScrollView>
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "space-evenly",
      }
    }
  >
    <View
      accessible={true}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      testID="delete-btn"
    >
      <Text />
    </View>
    <View
      accessible={true}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      testID="save-btn"
    >
      <Text />
    </View>
  </View>
</View>
`;
